### ISSUE TITLE
Virtual dependencies incorrectly affecting the hash

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#4492]
@tgamblin @scheibelp @alalazo @davydden This is a really weird one that I've never seen before. 

From a fresh clone of Spack, with no `packages.yaml`, I'm seeing the following behavior. If I don't specify an MPI provider and let it pick up the default (`openmpi`), I see:
```
$ spack spec -L hdf5+mpi
...
tgprd6nwfgifpaff4a7dvcv3tynuaowd  hdf5@1.10.1%gcc@6.3.1+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=linux-fedora25-x86_64 
hj3i67mhesgu3x5ymwpj67m4nbq6d3c2      ^openmpi@2.1.1%gcc@6.3.1~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple+vt arch=linux-fedora25-x86_64 
eva5cqpkbjggxvpp23ycr62yiocie64g          ^hwloc@1.11.7%gcc@6.3.1~cuda+libxml2+pci arch=linux-fedora25-x86_64 
reafmcemqbmvzqzyiitmyfcdc4ia7tzl              ^libpciaccess@0.13.5%gcc@6.3.1 arch=linux-fedora25-x86_64 
4w4hvmepes4p75z37og3dn3gbzjpnwho                  ^libtool@2.4.6%gcc@6.3.1 arch=linux-fedora25-x86_64 
zcwwrrt7od3rrhxb5du2iybu65jlrlbw                      ^m4@1.4.18%gcc@6.3.1+sigsegv arch=linux-fedora25-x86_64 
625n6uiwuqes737tignpxu3snoibyc43                          ^libsigsegv@2.11%gcc@6.3.1 arch=linux-fedora25-x86_64 
ywniw4sq7xd3fzz6c4c7jrhqj5htwobl                  ^pkg-config@0.29.2%gcc@6.3.1+internal_glib arch=linux-fedora25-x86_64 
lnv6kl5fucdneeh7h7ickta4vdis7u4v                  ^util-macros@1.19.1%gcc@6.3.1 arch=linux-fedora25-x86_64 
ggtx25lpnfriqn4n35fsgquiug6uju7t              ^libxml2@2.9.4%gcc@6.3.1~python arch=linux-fedora25-x86_64 
yw6ju7g3kyc27eonzrrvu73cdpiy5hzq                  ^xz@5.2.3%gcc@6.3.1 arch=linux-fedora25-x86_64 
jcukpuaeyykzu647duoufe2bdrgru5ko                  ^zlib@1.2.11%gcc@6.3.1+pic+shared arch=linux-fedora25-x86_64 
```
If I do specify the MPI provider and explicitly tell it to use `openmpi`, I see:
```
$ spack spec -L hdf5+mpi^openmpi
...
ohnyu6ch2ewnkf2dnd32qrb6t2hlopgj  hdf5@1.10.1%gcc@6.3.1+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=linux-fedora25-x86_64 
yyiri5gxf7jtu2c3vrhj6on4wz3jf2jh      ^openmpi@2.1.1%gcc@6.3.1~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple+vt arch=linux-fedora25-x86_64 
eva5cqpkbjggxvpp23ycr62yiocie64g          ^hwloc@1.11.7%gcc@6.3.1~cuda+libxml2+pci arch=linux-fedora25-x86_64 
reafmcemqbmvzqzyiitmyfcdc4ia7tzl              ^libpciaccess@0.13.5%gcc@6.3.1 arch=linux-fedora25-x86_64 
4w4hvmepes4p75z37og3dn3gbzjpnwho                  ^libtool@2.4.6%gcc@6.3.1 arch=linux-fedora25-x86_64 
zcwwrrt7od3rrhxb5du2iybu65jlrlbw                      ^m4@1.4.18%gcc@6.3.1+sigsegv arch=linux-fedora25-x86_64 
625n6uiwuqes737tignpxu3snoibyc43                          ^libsigsegv@2.11%gcc@6.3.1 arch=linux-fedora25-x86_64 
ywniw4sq7xd3fzz6c4c7jrhqj5htwobl                  ^pkg-config@0.29.2%gcc@6.3.1+internal_glib arch=linux-fedora25-x86_64 
lnv6kl5fucdneeh7h7ickta4vdis7u4v                  ^util-macros@1.19.1%gcc@6.3.1 arch=linux-fedora25-x86_64 
ggtx25lpnfriqn4n35fsgquiug6uju7t              ^libxml2@2.9.4%gcc@6.3.1~python arch=linux-fedora25-x86_64 
yw6ju7g3kyc27eonzrrvu73cdpiy5hzq                  ^xz@5.2.3%gcc@6.3.1 arch=linux-fedora25-x86_64 
jcukpuaeyykzu647duoufe2bdrgru5ko                  ^zlib@1.2.11%gcc@6.3.1+pic+shared arch=linux-fedora25-x86_64 
```
As you can see, the packages to be installed are identical. However, the hash for `openmpi` (and consequently `hdf5`) changed!

**Comment by adamjstewart** [comment]
And the plot thickens. `spack install` correctly detects that the package is already installed even though `spack spec -I` reports that it is not.

**Comment by alalazo** [comment]
Ouch, that hurts: `git bisect` tells me it's coming from #2386

**Comment by davydden** [comment]
could be ultimately the same bug somewhere in conretization as in https://github.com/LLNL/spack/issues/4107 but manifests itself differently.

**Comment by davydden** [comment]
p.s. I also noticed that `spack spec -I` sometimes incorrectly reports a package as not being installed.

**Comment by alalazo** [comment]
@tgamblin @adamjstewart I'll try to debug where this is coming from. Will post here anything I find.

**Comment by alalazo** [comment]
Debugging session running `spack spec -L hdf5+mpi` at line 92 of the command (after `spec.concretize()`):
```python
In[1]: openmpi = spec['openmpi']

In[2]: openmpi.dag_hash()
Out[3]: 'gw3622ptsqq5tuvxrd5z3xynjysldufi'

In[3]: openmpi._hash = None

In[4]: openmpi.dag_hash()
Out[5]: 'his36lswhij5wjcfll7rgprw4lt6sfvb'
```
In a session running `spack spec -L hdf5+mpi ^openmpi`:
```python
In[1]: openmpi = spec['openmpi']

In[2]: openmpi.dag_hash()
Out[3]: 'his36lswhij5wjcfll7rgprw4lt6sfvb'

In[3]: openmpi._hash = None

In[4]: openmpi.dag_hash()
Out[5]: 'his36lswhij5wjcfll7rgprw4lt6sfvb'
```
In the first case it seems we are caching the result of hashing too early, or wrongly in some way.

**Comment by alalazo** [comment]
Not sure why, but in the first case (`spack spec -L hdf5+mpi`) `openmpi` is marked concrete when dependencies are not yet added to it. We are taking the hash of the following:
```yaml
{openmpi: {version: 2.1.1, arch: {platform: linux, platform_os: ubuntu14, target: x86_64},
    compiler: {name: gcc, version: '4.8'}, namespace: builtin, parameters: {cuda: false,
      fabrics: [''], java: false, schedulers: [''], sqlite3: false, thread_multiple: false,
      vt: true, cflags: [], cppflags: [], cxxflags: [], fflags: [], ldflags: [], ldlibs: []}}}
```
In the second case we are hashing instead the correct yaml dump, which is:
```yaml
{openmpi: {version: 2.1.1, arch: {platform: linux, platform_os: ubuntu14, target: x86_64},
    compiler: {name: gcc, version: '4.8'}, namespace: builtin, parameters: {cuda: false,
      fabrics: [''], java: false, schedulers: [''], sqlite3: false, thread_multiple: false,
      vt: true, cflags: [], cppflags: [], cxxflags: [], fflags: [], ldflags: [], ldlibs: []},
    dependencies: {hwloc: {hash: e7pwcoccg3svl7lzzu5qqd65w25zhbvt, type: [build, link]}}}}
```

**Comment by alalazo** [comment]
**TLDR**: possibly found the bug. It's in:
```python
class DependencyMap(HashableMap):
    """Each spec has a DependencyMap containing specs for its dependencies.
       The DependencyMap is keyed by name. """

    @property
    def concrete(self):
        return all((d.spec.concrete and d.deptypes)
                   for d in self.values())
...
```
and it is due to the fact that `concrete` will return `True` for an empty set of values.

---

Longer explanation. The first command (`spack spec -L hdf5+mpi`) takes the wrong path at `spec.py:2290`, when evaluating:
```python
# If self is a concrete spec, and other is not virtual, then we need
# to substitute every multi-valued variant that needs it with a
# single-valued variant.
if self.concrete: # < -- Evaluating this line
    try:
        # When parsing a spec every variant of the form
        # 'foo=value' will be interpreted by default as a
        # multi-valued variant. During validation of the
        # variants we use the information in the package
        # to turn any variant that needs it to a single-valued
        # variant.
        substitute_single_valued_variants(other)
    except (SpecError, KeyError):
        # Catch the two things that could go wrong above:
        # 1. name is not a valid variant (KeyError)
        # 2. the variant is not validated (SpecError)
        return False
```
which was introduced as part of #2386 (and explains the bisect), because there's a possible bug in the way `self.concrete` is computed:
```python
@property
def concrete(self):
    """A spec is concrete if it can describe only ONE build of a package.
       If any of the name, version, architecture, compiler,
       variants, or depdenencies are ambiguous,then it is not concrete.
    """
    if self._concrete:
        return True

    self._concrete = bool(not self.virtual and
                          self.namespace is not None and
                          self.versions.concrete and
                          self.variants.concrete and
                          self.architecture and
                          self.architecture.concrete and
                          self.compiler and self.compiler.concrete and
                          self.compiler_flags.concrete and
                          self._dependencies.concrete) # <-- The problem is at this line
    return self._concrete
```

When we are evaluating `self.concrete` in this context we are in the middle of concretization, and `openmpi` has no dependencies attached but is otherwise concrete. The method to check if the dependencies are concrete reads:
```python
@property
def concrete(self):
    return all((d.spec.concrete and d.deptypes)
               for d in self.values())
```
which is the same as:
```python
In[3]: all(tuple())
Out[4]: True
```
and gives the wrong result as `openmpi` has no dependencies attached yet.

**Comment by davydden** [comment]
@alalazo I guess one can not simply check in `DependencyMap` if `self.values()` is empty and return `False` (not-conrete) as this may be a package without any dependencies. Probably one has to add additional flag somewhere to say if a package has dependencies. Or maybe modify the logic in `self._concrete` of Spec.

**Comment by alalazo** [comment]
@davydden I was thinking of adding a constructor to `DependencyMap` similar to the one in `FlagMap`:
```python
def __init__(self, spec):
    super(FlagMap, self).__init__()
    self.spec = spec
```
Once we have the reference to the `spec` we may query the corresponding package to see:

1. If we satisfy every dependency in the directives that are "active" (meaning the spec satisfies their condition)
2. If the dependencies are concrete and of the correct type

The latter point is already coded in the `DependencyMap.concrete` property. The former is what is missing. I'll try to put together a PR tomorrow. Let me know if you have any suggestion.

**Comment by olesenm** [comment]
@alalazo The intermediate resolving seems to be odd as well. Testing the openfoam virtual package.

    $ spack spec openfoam@plus
    Input spec:  openfoam@plus
    Normalized: openfoam@plus
    Concretized: openfoam-com@1612%gcc@4.8 ...

Whereas 

    $ spack spec openfoam-com@plus
    Input spec:  openfoam-com@plus
    Normalized: openfoam-com@plus
    Concretized: openfoam-com@plus%gcc@4.8 ...

version information is ignored when a virtual is specified?

**Comment by alalazo** [comment]
@olesenm I don't think what you see relates to this bug. Virtual dependencies *have their own version*, in the sense that the version you attach to a virtual dependency represents the API version. To make a concrete example `mpi@3` means that I want the standard 3 of MPI, not that I want to install `openmpi@3` (assuming the provider chosen by the concretizer is `openmpi`).

**Comment by olesenm** [comment]
OK - thanks for the explanation (makes sense). I guess that I probably get the '1612' since it sort before 'plus'.
